File: src/LeetCode/ArrayAndString/ReverseString.java
package LeetCode.ArrayAndString;

import java.util.Arrays;

public class ReverseString {
    final private char[] s;

    public ReverseString(char[] s) {
        this.s = s;
    }

    public void swap(char[] s, int left, int right) {
        char temp = s[left];

        s[left] = s[right];
        s[right] = temp;
    }


    // Time complexity: O(n)
    // Space complexity: O(1)
    public void naive() {
        int left = 0;
        int right = s.length - 1;


        while (left < right) {
            swap(s, left, right);

            left++;
            right--;
        }

        System.out.printf(Arrays.toString(s));
    }

}

File: dsaLeetCode/arrayAndString/twoPointers/reverseString.js
// Time Complexity: O(n)
// Space Complexity: O(1)
const reverseString = (s) => {
  let leftIdx = 0;
  let rightIdx = s.length - 1;

  while(leftIdx < rightIdx) {
    [s[leftIdx], s[rightIdx]] = [s[rightIdx], s[leftIdx]];

    leftIdx++;
    rightIdx--;
  };

  return s;
};

console.log(reverseString(["h", "e", "l", "l", "o"]));
